channels/generic/http.py

```diff
@@ -80,10 +100,15 @@ def __init__(self, *args, **kwargs):
         if not message.get("more_body"):
             try:
                 await self.handle(b"".join(self.body))
-            finally:
-                await self.disconnect()
-                raise StopConsumer()
+            except StopConsumer:
+                await self.close(status=200)
+            except:
+                # TODO This is just a patch, after bubbling up the exception no body is calling http_disconnect.
+                await self.close(body=b"Internal Server Error", status=500)
```

Comment:
I wander if it would be nicer to have a `exception_handler` method on the `AsyncHttpConsumer` that sends this 500 down the wire. 

In some cases we need to format this 500 response a little differently (eg, if we got a Redis connection exception we might well want to push some retry after headers on the response)

Reply:
Interesting... if you want  to format it differently you can always catch the exception in your class and handle it in your code.
Honestly I'm not familiar enough with the code base, I think when any exception bubbles up the http connection should be closed and the consumer stopped, but that doesn't happen for some reason. So this is the fix that I found.

Reply:
btw ... be sure to re-raise `asyncio.CancelledError`. If you don't re-raise it, asyncio produces undefined behavior.

For that matter ... shouldn't _every_ exception be re-raised? I'm a keen user of `asyncio.EventLoop.set_exception_handler()` -- it helps find bugs on production. (I imagine this is why `finally` was used in the first place.)

Reply:
I agree un-expected exceptions should not be swallowed so catching in `self.handle` and sending a response would not be ideal. 

the reason for wanted to be able to manage the 500 response body is that for our server we want to return a JSON error bodys that is parsable. At the moment our solution for this is to put a middleware layer between the Consume and the top level Application that proxies all `send` messages however since this does not have access to the source exception it can't produce more than a very generic error.

@adamhooper The issue with the `finally` approach currently in master is that it replaces effectively swallows your exception, since finally raises `StopConsumer` any exception that was raised is then dropped by the AsyncConsume parent classes `__call__` method.

Reply:
@carltongibson what do we think about:`set_exception_handler` or `CancelledError` being used here? That/and/or having us create an `exception_handler` method to keep the flow more organized?

Are any of the pieces in this discussion required vs optional, or is the flow a bit beyond straightforward rule-logic?

Reply:
My thought on this, when I looked at it, was that we're entering a loosing battle. (I catch an exception, try and call close, and end awaiting body — but hang-on! I was closing.) 

Rather, for this kind of _something I really wasn't expecting went wrong_ kind of case, I think we should just let the exception bubble up to Daphne, [where it will be handled](https://github.com/django/daphne/blob/59b57a9f4b7a419f1ab3ecb792ca802671045075/daphne/http_protocol.py#L287-L291) with a plain response (and perhaps we can add some logging to help the related issues around this space...). 

Then, assuming that all looks good, we need to make sure Daphne sends the appropriate `http_disconnect` to stop the consumer. (Which looks like it's not happening.) 

(Of course, this isn't fresh in my mind so I may need to fire up the debugger again to get back on top of it.)